src/displayBackend/common/ConnectorBase.cpp

```diff
@@ -19,12 +19,329 @@
  *
  */
 
+#include <xen/be/Exception.hpp>
+
 #include "ConnectorBase.hpp"
 
+#include <byteswap.h>
+#include "drm_edid.h"
+
+using XenBackend::Exception;
+using XenBackend::XenGnttabBuffer;
+
 ConnectorBase::ConnectorBase(domid_t domId, uint32_t width, uint32_t height) :
 	mDomId(domId),
 	mCfgWidth(width),
 	mCfgHeight(height),
 	mLog("Connector")
 {
 }
+
+/*******************************************************************************
+ * Protected
+ ******************************************************************************/
+
+#if (XENDISPL_PROTOCOL_VERSION >= 2)
+
+uint32_t ConnectorBase::toLittleEndian32(uint32_t val)
+{
+#ifdef IS_BIG_ENDIAN
+	return __bswap_32(val);
+#else
+	return val;
+#endif
+}
+
+void ConnectorBase::edidPutBlockCheckSum(uint8_t* edidBlock)
+{
+	int i, checkSum = 0;
+
+	for (i = 0; i < XENDISPL_EDID_BLOCK_SIZE - 1; i++)
+	{
+		checkSum += edidBlock[i];
+	}
+
+	edidBlock[i] = 0x100 - static_cast<uint8_t>(checkSum);
+}
+
+void ConnectorBase::edidPutEssentials(edid* edidBlock)
+{
+	const uint8_t edidHeader[] = {
+		0x00, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00
+	};
+
+	memcpy(&edidBlock->header, edidHeader, sizeof(edidHeader));
+
+	/* 16-bit value made up of three 5-bit letters, big-endian. */
+	uint16_t vendorId = (((('X' - '@') & 0x1f) << 10) |
+						 ((('E' - '@') & 0x1f) <<  5) |
+						 ((('N' - '@') & 0x1f) <<  0));
+#ifdef IS_BIG_ENDIAN
+	edidBlock->mfg_id[0] = vendorId & 0xff;
+	edidBlock->mfg_id[1] = vendorId >> 8;
+#else
+	edidBlock->mfg_id[1] = vendorId >> 8;
+	edidBlock->mfg_id[0] = vendorId & 0xff;
+#endif
+
+	/* Manufacturer product code. 16-bit number, little-endian. */
+#ifdef IS_BIG_ENDIAN
+	edidBlock->prod_code[0] = 'M';
+	edidBlock->prod_code[1] = 'V';
+#else
+	edidBlock->prod_code[0] = 'M';
+	edidBlock->prod_code[1] = 'V';
+#endif
+
+	/* Serial number. 32 bits, little-endian. */
+	edidBlock->serial = toLittleEndian32(0x20200505);
+
+	/* Week of manufacture - not consistent between manufacturers. */
+	edidBlock->mfg_week = 1;
+	/* Year of manufacture. Year = datavalue + 1990. */
+	edidBlock->mfg_year = 2020 - 1990;
+
+	/* EDID version and revision: advertise 1.4. */
+	edidBlock->version = 1;
+	edidBlock->revision = 4;
+
+	/*
+	 * Video input parameters bitmap: digital, 8 bits per color,
+	 * display port.
+	 */
+	edidBlock->input = DRM_EDID_INPUT_DIGITAL | DRM_EDID_DIGITAL_DEPTH_8 |
+		DRM_EDID_DIGITAL_TYPE_DP;
+
+	/* Set width and hight as undefined. */
+	edidBlock->width_cm = 0;
+	edidBlock->height_cm = 0;
+
+	/* Display gamma, 2.2. */
+	edidBlock->gamma = 220 - 100;
```

Comment:
Fixing this value imply de-gamma & gamma process in case if display gamma of backend DRM device is different. Not all HW can do this which may require SW operation -> consume extra CPU power

Reply:
will fix